src/jit/gentree.cpp

```diff
@@ -6122,21 +6134,29 @@ GenTreePtr Compiler::gtNewIconEmbHndNode(void* value, void* pValue, unsigned fla
 {
     GenTreePtr node;
 
-    assert((!value) != (!pValue));
-
-    if (value)
+    if (value != nullptr)
     {
+        // When 'value' is non-null, pValue is required to be null
+        assert(pValue == nullptr);
+
+        // use 'value' to construct an integer constant node
         node = gtNewIconHandleNode((size_t)value, flags, /*fieldSeq*/ FieldSeqStore::NotAField());
         node->gtIntCon.gtCompileTimeHandle = (size_t)compileTimeHandle;
     }
     else
     {
-        node = gtNewIconHandleNode((size_t)pValue, flags, /*fieldSeq*/ FieldSeqStore::NotAField());
-        node->gtIntCon.gtCompileTimeHandle = (size_t)compileTimeHandle;
-        node                               = gtNewOperNode(GT_IND, TYP_I_IMPL, node);
+        // When 'value' is null, pValue is required to be non-null
+        assert(pValue != nullptr);
+
+        // use 'pValue' to construct an integer constant node
+        GenTreePtr addrNode = gtNewIconHandleNode((size_t)pValue, flags);
+        addrNode->gtIntCon.gtCompileTimeHandle = (size_t)compileTimeHandle;
 
-        // This indirection won't cause an exception.  It should also
-        // be invariant, but marking it as such leads to bad diffs.
+        // construct the indirection of 'pValue'
+        node = gtNewOperNode(GT_IND, TYP_I_IMPL, addrNode);
+
+        // This indirection won't cause an exception.  It should also be invariant,
+        // but marking it as such current leads to lots of extra CSEs resulting in bad diffs.
```

Comment:
Why don't you use your new gtNewIndOfIconHandleNode? (it appears you removed it)

Reply:
I am not using gtNewIndOfIconHandleNode here because there is an assignment to gtCompileTimeHandle in this sequence and when using gtNewIndOfIconHandleNode  you don't have a direct handle to the address constant.